docs: Consul K8s Overview update (#12575)
hashicorp/consul: website/content/docs/k8s/architecture.mdx

---
layout: docs
page_title: Consul on Kubernetes Architecture
description: >-
 A high level overview of Consul on Kubernetes Architecture 
---

# Architecture

> Consul runs on Kubernetes with the same
> [architecture](/docs/architecture)
> as other platforms. There are some benefits Kubernetes can provide
> that eases operating a Consul cluster and we document those below. The standard
> [production deployment guide](https://learn.hashicorp.com/consul/datacenter-deploy/deployment-guide) is still an
> important read even if running Consul within Kubernetes.
> 
> Each section below will outline the different components of running Consul
> on Kubernetes and an overview of the resources that are used within the
> Kubernetes cluster.

Reviewer: ```suggestion
This topic describes the architecture, components, and resources associated with Consul deployments to Kubernetes. Consul employs the same architectural design on Kubernetes as it does with other platforms (see [Architecture](/docs/architecture)), but Kubernetes provides additional benefits that make operating a Consul cluster easier. 

Refer to the standard [production deployment guide](https://learn.hashicorp.com/consul/datacenter-deploy/deployment-guide) for important information, regardless of the deployment platform.
```
I just want to be clear about what this topic covers.

## Server Agents

> The server agents are run as a **StatefulSet**, using persistent volume

Reviewer: ```suggestion
The server agents are run as a **StatefulSet**, using persistent volume
```
I think that because the Kubernetes keywords are configuration settings (or appear to be based on my Google search), we should use code font and reposition how we reference them to make this clear. 

For example, we might want to rephrase this sentence to something like:

Consul server agents operate as `StatefulSets` types and use persistent . . . "

More precise, no?

Author: Thanks!

claims to store the server state. This also ensures that the
[node ID](/docs/agent/options#_node_id) is persisted so that servers
can be rescheduled onto new IP addresses without causing issues. The server agents
are configured with
[anti-affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity)
rules so that they are placed on different nodes. A readiness probe is
configured that marks the pod as ready only when it has established a leader.

> A **Service** is registered to represent the servers and expose the various
> ports. The DNS address of this service is used to join the servers to each
> other without requiring any other access to the Kubernetes cluster. The
> service is configured to publish non-ready endpoints so that it can be used
> for joining during bootstrap and upgrades.

Reviewer: ```suggestion
Configure and register a Kubernetes `Service` to represent the servers and expose the various
ports. The service's DNS address joins the servers to each other without requiring access to the Kubernetes cluster. You should configure the service to publish endpoints that may not be ready so that the endpoints can join servers during bootstrap and upgrades.
```
I think that one of the confusing bits (for me, anyway) is that this is written in passive voice. It sounds as if this all happens automatically when you install Consul. Even if the operations are performed programmatically without a clear "you", we should try to use active/direct language so it's clear that we are directing users to do something.

Also, I think this suggestion retains the meaning, but I was a little unclear about that last sentence.

> Additionally, a **PodDisruptionBudget** is configured so the Consul server

Reviewer: See previous comment about passive voice and code font for K8s entities.
